src/crimson/os/seastore/btree/fixed_kv_node.h

```diff
@@ -28,231 +28,32 @@ namespace crimson::os::seastore {
-      children(rhs.capacity, nullptr),
-      capacity(rhs.capacity) {}
+    : CachedExtent(rhs),
+      range(rhs.range) {}
 
   virtual fixed_kv_node_meta_t<node_key_t> get_node_meta() const = 0;
-  virtual uint16_t get_node_size() const = 0;
-
   virtual ~FixedKVNode() = default;
-  virtual node_key_t get_key_from_idx(uint16_t idx) const = 0;
-
-  template<typename iter_t>
-  void update_child_ptr(iter_t iter, ChildableCachedExtent* child) {
-    children[iter.get_offset()] = child;
-    set_child_ptracker(child);
-  }
-
-  virtual bool is_leaf_and_has_children() const = 0;
-
-  template<typename iter_t>
-  void insert_child_ptr(iter_t iter, ChildableCachedExtent* child) {
-    auto raw_children = children.data();
-    auto offset = iter.get_offset();
-    std::memmove(
-      &raw_children[offset + 1],
-      &raw_children[offset],
-      (get_node_size() - offset) * sizeof(ChildableCachedExtent*));
-    if (child) {
-      children[offset] = child;
-      set_child_ptracker(child);
-    } else {
-      // this can happen when reserving lba spaces and cloning mappings
-      ceph_assert(is_leaf_and_has_children());
-      // this is to avoid mistakenly copying pointers from
-      // copy sources when committing this lba node, because
-      // we rely on pointers' "nullness" to avoid copying
-      // pointers for updated values
-      children[offset] = get_reserved_ptr();
-    }
-  }
-
-  template<typename iter_t>
-  void remove_child_ptr(iter_t iter) {
-    LOG_PREFIX(FixedKVNode::remove_child_ptr);
-    auto raw_children = children.data();
-    auto offset = iter.get_offset();
-    SUBTRACE(seastore_fixedkv_tree, "trans.{}, pos {}, total size {}, extent {}",
-      this->pending_for_transaction,
-      offset,
-      get_node_size(),
-      (void*)raw_children[offset]);
-    // parent tracker of the child being removed will be
-    // reset when the child is invalidated, so no need to
-    // reset it here
-    std::memmove(
-      &raw_children[offset],
-      &raw_children[offset + 1],
-      (get_node_size() - offset - 1) * sizeof(ChildableCachedExtent*));
-  }
-
-  virtual bool have_children() const = 0;
+  virtual void _on_rewrite(Transaction &t, CachedExtent &extent) = 0;
 
   void on_rewrite(Transaction &t, CachedExtent &extent, extent_len_t off) final {
     assert(get_type() == extent.get_type());
     assert(off == 0);
-    auto &foreign_extent = (FixedKVNode&)extent;
     range = get_node_meta();
-
-    if (have_children()) {
-      if (!foreign_extent.is_pending()) {
-	foreign_extent.add_copy_dest(t, this);
-	copy_sources.emplace(&foreign_extent);
-      } else {
-	ceph_assert(foreign_extent.is_mutation_pending());
-	auto copy_source = 
-	  foreign_extent.get_prior_instance()->template cast<FixedKVNode>();
-	copy_source->add_copy_dest(t, this);
-	copy_sources.emplace(copy_source);
-	children = std::move(foreign_extent.children);
-	adjust_ptracker_for_children();
-      }
-    }
+    _on_rewrite(t, extent);
```

Comment:
Let's be consistent about the style about where might be extended for parent-to-child ptr, to make the extension places explicit and easier to follow.

In this example: `on_rewrite()` needs to be extended, so introduce a `do_on_rewrite()`, with the original interface finalized.

Same to `on_replace_prior()`, `prepare_commit()`, `on_clean_read()`, `apply_delta_and_adjust_crc()`. They are based on overriding instead, I feel it's easy to get lost to identify the exact call path, as there are multiple levels of complex inheritances.

What do you think?

Reply:
Yeah, will do.